decompiler/structures/pseudo/z3_logic.py

```diff
@@ -0,0 +1,216 @@
+            return self._convert_constant(expression)
+        if isinstance(expression, Branch):
+            return self._convert_branch(expression)
+        if isinstance(expression, Condition):
+            return self._convert_condition(expression)
+        if isinstance(expression, Operation):
+            return self._convert_operation(expression)
+        raise ValueError(f"Could not convert {expression} into z3 logic.")
+
+    def _convert_variable(self, variable: Variable) -> BitVecRef:
+        """Represent the given Variable as a BitVector in z3."""
+        return BitVec(variable.name, variable.type.size if variable.type.size else 32, ctx=self._context)
+
+    def _convert_constant(self, constant: Constant) -> BitVecRef:
+        """Represent the given variable as a BitVector (no types)."""
+        return BitVecVal(constant.value, constant.type.size if constant.type.size else 32, ctx=self._context)
+
+    def _convert_branch(self, branch: Branch) -> BitVecRef:
+        """
+        Convert the given branch into z3 logic.
+
+        Condition type Branches can be converted as a BinaryOperation.
+        When the condition is a single variable or an expression, the condition becomes a check != 0.
+        """
+        if isinstance(branch.condition, Condition):
+            return self._convert_condition(branch.condition)
+        return self._convert_condition(Condition(OperationType.not_equal, [branch.condition, Constant(0, branch.condition.type)]))
+
+    def _convert_condition(self, condition: Condition) -> BitVecRef:
+        """
+        Convert the given condition into z3 logic.
+
+        Please note conditions are a special kind of operation returning a boolean value.
+        """
+        _operation = self._get_operation(condition)
+        return self._ensure_bool_sort(_operation)
+
+    def _convert_operation(self, operation: Operation) -> BitVecRef:
+        """
+        Convert the given operation into a z3 logic.
+
+        Operations should always be of BitvecRef type.
+        """
+        _operation = self._get_operation(operation)
+        return self._ensure_bitvec_sort(_operation)
+
+    def _get_operation(self, operation: Operation) -> Union[BoolRef, BitVecRef]:
+        """Convert the given operation into a z3 expression utilizing the handler functions."""
+        operands = self._ensure_same_sort([self.convert(operand) for operand in operation.operands])
+        if isinstance(operands[0], BoolRef) and operation.operation in self.OPERATIONS_BOOLREF:
+            converter = self.OPERATIONS_BOOLREF.get(operation.operation, None)
+        else:
+            converter = self.OPERATIONS.get(operation.operation, None)
+        if not converter:
+            raise ValueError(f"Could not convert operation {operation} into z3 logic.")
+        return converter(*operands)
+
+    def _ensure_same_sort(self, operands: List[ExprRef]) -> List[ExprRef]:
```

Comment:
Before I forget it later, I am currently fixing a bug regarding this function #33 #75. The change should be this function and a test.

Reply:
Looks like the fix is already on main. Will rebase on top later!

Reply:
Rebased.